### PR TITLE
Introducing types and enforcing checks as part of the CI

### PR DESCRIPTION
**Pull request by acabarbaye**
### Description
Use of python3 types in project source
Use of `mypy` to check their use in the project

This PR depends on #3 
### Test Coverage

- [ ]  This change is covered by existing or additional automated tests.
- [ ]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [x]  Additional tests are not required for this change (e.g. documentation update).

### CODE REVIEW

File: mbed_targets/mbed_targets.py [hunk 1]
```diff
@@ -1,53 +1,53 @@
...
-        self._features = self._attributes.get("features", {})
+        self._target_entry: dict = target_database_entry
+        self._attributes: dict = self._target_entry.get("attributes", {})
+        self._features: dict = self._attributes.get("features", {})
 
     @property
-    def board_type(self):
+    def board_type(self) -> str:
         """Compilation target.
         :rtype: str
```
Comment (madchutney) [inline mbed_targets/mbed_targets.py]: I think this kind of comment now superfluous?

File: Pipfile [hunk 2]
```diff
@@ -6,9 +6,9 @@ verify_ssl = true
 [dev-packages]
 mock = "*"
 requests-mock = "*"
-jinja2 = "*"
```
Comment (rosiemiller2) [inline Pipfile]: How come jinja2 isn't needed anymore?

File: azure-pipelines/build-release.yml [hunk 3]
```diff
@@ -62,7 +62,11 @@ stages:
     # Static analysis is different for Python versions so should be run in each environment.
     - script: |
         flake8
-      displayName: 'Static Analysis (flake8)'
+      displayName: 'Static Analysis - general (flake8)'
+
+    - script: |
+        mypy -p mbed_targets
+      displayName: 'Static Analysis - type checks (mypy)'
```
Comment (rosiemiller2) [inline azure-pipelines/build-release.yml]: Something a little off with the indentation here.

File: Pipfile (hunk 2 above)
Comment (acabarbaye) [reply Pipfile]: jinja2 is used by the target data inconsistencies report which is in /ci_scripts and is not needed by the actual mbed-target code

**Pull request merged**